Fix is_positive for gamma and floor & ceiling is_even Bug.

aktech opened the pull request:
The Positive assumption helper for gamma function is missing:
For Reference (Proof) : (See graph of Gamma Function)
http://mathworld.wolfram.com/GammaFunction.html

Gamma Function is positive for: 
- Positive value of x & 
- Alternately unit ranges between -ve integers where the floor of x is even. (see graph in link above)

&
This should return False.

``` python
In [7]: x =Symbol('x', even=True, negative=True)

In [8]: floor(x - S.Half).is_even
Out[8]: True
```

Fixes #8524
closes #8864
@smichr  Please have a look.

smichr commented on sympy/functions/special/gamma_functions.py:
```diff
@@ -160,6 +160,13 @@ def _eval_is_real(self):
+        x = self.args[0]
+        if x.is_positive:
+            return True
+        elif x.is_noninteger:
```
should we add `elif x.is_nonpositive and x.is_integer: return False`? Otherwise this looks ok but could use a test like 
```
assert gamma(Symbol('e', even=True, negative=True) + S.Half).is_positive is True
assert gamma(Symbol('e', even=True, negative=True) - S.Half).is_positive is False
```

aktech replied:
Thanks! , that would be good, I will add them, and the second test is expected to Fail as our assumption system have this issue:
`floor(x - S.Half).is_even` returns `True` for negative even 'x' .
I have reported the issue at [here](https://github.com/sympy/sympy/issues/8853) .
Anyways we can XFAIL that.

aktech replied:
@smichr 
Can you tell me:
What's the best way to add XFAIL within a test/function,
Example:
```python
def test_foo_bar():
    assert something1
    assert something2    #this line is expected to fail, how to add XFAIL for this?
    assert something3
```

smichr replied:
```
@XFAIL
def test_foo_bar_fail():
    assert something2
```

No shortcut other than factoring it out, AFAIK.

aktech replied:
Ok, thanks!
We will not require this after merging a PR for #8853

smichr commented:
OK, this can be rebased over #8864 and the tests can be added.

aktech commented:
@smichr 
Thanks, I will add tests when #8864 gets merged, so that test doesn't fails for this PR.

smichr commented:
Could you pull my evalf branch and rebase this over that and force push it here again? There was another change that needed to be made. Then the other PR can just be closed (since tests will fail) and this one will contain the work of both PRs. If you want to wait until tests pass [here](https://travis-ci.org/smichr/sympy/builds/47805947), however, that would be best.

smichr commented:
(After doing the above then you should be able to add the tests that we talked about in this PR.)

smichr commented:
in commit title: ceilling -> ceiling

aktech commented:
I have squashed all commits to one, shall I push that?

smichr commented:
sure...let's see what that looks like.

aktech commented:
I accidentally messed with that, let me fix that.

aktech commented:
I have squashed all commits in to one.

smichr committed:
watch for all rationals in calc_part

The number reresented with n, c, p in calc_part is [1, -1][n]*c*2**p.
with n,c,p all integers. A non-integer will have p < 0 so that test was
added to the existing test. Th c!=1 and p!=0 test is for a computed
integer; something that was a pure integer would have c=1 or p=0.

aktech committed:
Fix is_positive for gamma

smichr commented:
Hmm. I would prefer to keep the changes separate since I included some commit notes. I created a new branch evalf which if you check it out, delete your gammap branch, and then rename it (the evalf branch that you are in) to gammap and force-push to this PR everything should be fine. Just be sure not to close this PR while doing that.

BTW, if you look at the tests I added you will see two groups: in the one I am testing something that should always be true -- floor and ceiling behavior -- and in the 2nd group I am testing the specific thing that was broken. That way even if someone deletes the calc_part function we still guarantee that the desired behavior is not lost.

The [tests](https://travis-ci.org/smichr/sympy/builds/47805947) are looking good except that that run contained a print statement that has been removed so I think we are done iterating on this. I'll commit everything in about 18 hrs.

aktech commented:
Thanks! for fixing the mess around commits.
Everything looks good to me, except for this
```
p = Symbol('x', even=True, positive=True)
```
I was wondering, why assign `x` to Symbol `p`?
wouldn't this be better `p = Symbol('p', even=True, positive=True)` ?

smichr commented:
doh...that's why I tend to just use `Dummy(even=True, positive=True)` in such circumstances...because I'm a dummy! LOL Let's just leave it -- there doesn't have to be a correspondence between name and variable name and since there are no other symbols in the test there is no confusion. (But if you want to go back and edit that commit I'm fine with that.)

aktech commented:
OK thanks!
I thought if there was any other significance. It's clear to me now, I am OK with it.

smichr commented:
We've already seen the tests pass in all versions and the test [here](https://travis-ci.org/smichr/sympy/builds/47843611) is good. So this is launched...glad to see that floor/ceiling bug taken care of, too. This is a constant theme of the work -- solve one, find one (or more)! Thanks for working on this.

aktech commented:
Thanks! for Lots of learning.
